File: fileMerging.py
# fileMerging.py | Merges two 3D images of same brain acquired from
# opposite directions, based on threshold values obtained from the
# "edge detection" step.
# Please see the user manual for details.
#
# Copyright (C) 2015, Dimitri Perrin
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.


from time import localtime, strftime
from subprocess import call
import ast
import sys


if len(sys.argv)!=2:
    print "\nUsage: "+sys.argv[0]+" <parameter_file>"
    quit()



# Reading the parameters

parameter_file = open(sys.argv[1],'r')
parameters = []
for line in parameter_file:
    if line[0] == "#":
        continue
    parameters.append(line.rstrip())
parameter_file.close()



# Processing the parameters

nii_dir = parameters[0]
ascii_dir = parameters[1]
merged_dir = parameters[2]
reg_dir = parameters[3]


brains = parameters[4].split(",")

sag_size = ast.literal_eval(parameters[5])
cor_size = ast.literal_eval(parameters[6])

hor_size = []
temp = parameters[7].split(",")
for i in range(0,len(temp)):
    hor_size.append(ast.literal_eval(temp[i]))

n_stain = []
temp = parameters[8].split(",")
for i in range(0,len(temp)):
    n_stain.append(ast.literal_eval(temp[i]))

m_stain = []
temp = parameters[9].split(",")
for i in range(0,len(temp)):
    m_stain.append(ast.literal_eval(temp[i]))

n_fluo = []
temp = parameters[10].split(",")
for i in range(0,len(temp)):
    n_fluo.append(ast.literal_eval(temp[i]))

m_fluo = []
temp = parameters[11].split(",")
for i in range(0,len(temp)):
    m_fluo.append(ast.literal_eval(temp[i]))

log = parameters[12]



# Checking the parameters

print "The method will: "
print " - read NIfTI-1 from "+nii_dir
print " - read ASCII from "+ascii_dir
print " - save registration data to "+reg_dir
print " - save merged NIfTI-1 to "+merged_dir
print " - save the log data to "+log

print "\nBrains to be processed: "
print brains



print " "
while 1:
    feedback = raw_input("Is this correct? (yes/no)\t").rstrip()
    if feedback == "yes":
        print "Program starting...\n"
        break
    if feedback == "no":
        print "Please edit the parameter file."
        quit()




# Running the analysis step


log_file = open(log,'w')



for i in range(0,len(brains)):
    output_message = strftime("%H:%M:%S", localtime())+": Starting to process brain "+brains[i]
    print output_message
    log_file.write(output_message+"\n")
    
    DV_stain = nii_dir+"DV_"+brains[i]+"_nuclear_aligned_VD.nii.gz"
    VD_stain = nii_dir+"VD_"+brains[i]+"_nuclear.nii.gz"
    DV_fluo = nii_dir+"DV_"+brains[i]+"_geneExp_aligned_VD.nii.gz"
    VD_fluo = nii_dir+"VD_"+brains[i]+"_geneExp.nii.gz"

    ascii_DV_stain = ascii_dir+"DV_"+brains[i]+"_nuclear.txt"
    ascii_VD_stain = ascii_dir+"VD_"+brains[i]+"_nuclear.txt"
    ascii_DV_fluo = ascii_dir+"DV_"+brains[i]+"_geneExp.txt"
    ascii_VD_fluo = ascii_dir+"VD_"+brains[i]+"_geneExp.txt"
    
    ascii_merged_stain = ascii_dir+"merged_"+brains[i]+"_nuclear.txt"
    ascii_merged_fluo = ascii_dir+"merged_"+brains[i]+"_geneExp.txt"
    
    merged_stain = merged_dir+brains[i]+"_nuclear.nii.gz"
    merged_fluo = merged_dir+brains[i]+"_geneExp.nii.gz"

    temp_stain_1 = merged_dir+"temp_nuclear_1.nii.gz"
    temp_fluo_1 = merged_dir+"temp_geneExp_1.nii.gz"
    temp_stain_2 = merged_dir+"temp_nuclear_2.nii.gz"
    temp_fluo_2 = merged_dir+"temp_geneExp_2.nii.gz"
    
    
    # Edge-based C-implemented file merging
    
    output_message = strftime("%H:%M:%S", localtime())+": Merging ASCII files"
    print output_message
    log_file.write(output_message+"\n")

    cmd = "./file_merging "+str(hor_size[i])+" "+str(sag_size)+" "+str(cor_size)+" "+ascii_DV_stain+" "+ascii_VD_stain+" "+ascii_merged_stain+" "+str(n_stain[i])+" "+str(m_stain[i])
    call([cmd],shell=True)
    
    cmd = "./file_merging "+str(hor_size[i])+" "+str(sag_size)+" "+str(cor_size)+" "+ascii_DV_fluo+" "+ascii_VD_fluo+" "+ascii_merged_fluo+" "+str(n_fluo[i])+" "+str(m_fluo[i])
    call([cmd],shell=True)
    
    
    # Converting from ASCII to NIfTI-1

    output_message = strftime("%H:%M:%S", localtime())+": Converting merged file to NIfTI-1"
    print output_message
    log_file.write(output_message+"\n")

    cmd = "fslascii2img "+ascii_merged_stain+" "+str(sag_size)+" "+str(cor_size)+" "+str(hor_size[i])+" 1 0.0258 0.0258 0.04 1 "+temp_stain_1
    call([cmd],shell=True)
    
    cmd = "fslascii2img "+ascii_merged_fluo+" "+str(sag_size)+" "+str(cor_size)+" "+str(hor_size[i])+" 1 0.0258 0.0258 0.04 1 "+temp_fluo_1
    call([cmd],shell=True)
    
    
    # Re-orienting the merged NIfTI-1 file
   
    output_message = strftime("%H:%M:%S", localtime())+": Reorienting"
    print output_message
    log_file.write(output_message+"\n")

    cmd = "c3d "+temp_stain_1+" -orient RAI -o "+temp_stain_2+" && rm "+temp_stain_1
    call([cmd],shell=True)
    
    cmd = "c3d "+temp_fluo_1+" -orient RAI -o "+temp_fluo_2+" && rm "+temp_fluo_1
    call([cmd],shell=True)
    

    # Registration to original VD_stain
    
    output_message = strftime("%H:%M:%S", localtime())+": Registration"
    print output_message
    log_file.write(output_message+"\n")

    reg = reg_dir+"merged_reg_"+brains[i]+"_.nii"
    log = reg_dir+"log_merged_"+brains[i]+".txt"

    cmd = "ANTS 3 -i 0 -o "+reg+" --MI-option 64x300000 -m CC["+VD_stain+","+temp_stain_2+",1,5] >> "+log
    call([cmd],shell=True)
    

    # Alignment

    output_message = strftime("%H:%M:%S", localtime())+": Alignment"
    print output_message
    log_file.write(output_message+"\n")
    
    cmd = "WarpImageMultiTransform 3 "+temp_stain_2+" "+merged_stain+" -R "+VD_stain+" "+reg[:-4]+"Affine.txt"
    call([cmd],shell=True)
    cmd = "WarpImageMultiTransform 3 "+temp_fluo_2+" "+merged_fluo+" -R "+VD_stain+" "+reg[:-4]+"Affine.txt"
    call([cmd],shell=True)
    cmd = "rm "+temp_stain_2+" "+temp_fluo_2
    call([cmd],shell=True)


output_message = strftime("%H:%M:%S", localtime())+": Done."
print output_message
log_file.write(output_message+"\n")

log_file.close()

File: file_merging.cpp
/*************************************************************************
 
 file_merging.cpp | Merges two 3D brain images given as ASCII files based
 on the threshold values calculated during "edge detection".
 Called from fileMerging.py
 Please see the user manual for details.
 
 Copyright (C) 2015, Dimitri Perrin
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 ************************************************************************/


#include <string>
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <iterator>

#define BUFFER_SIZE 10000

using namespace std;



/**************************************/
/**         printTimestamp()         **/
/**************************************/

/*
 Prints a custom string with the current time and a given message
 Input: the message we want to include
 Output: none
 */

int printTimestamp(string infoString) {
    time_t rawtime;
    struct tm * timeinfo;
    char buffer1 [15];
    
    time (&rawtime);
    timeinfo = localtime (&rawtime);
    strftime (buffer1,80,"%H:%M:%S ->\t",timeinfo);
    
    cout << buffer1 << infoString << endl;
    
    return 0;
}




/********************************/
/**         factor_DV()        **/
/********************************/

/*
 Calculates the weight for the DV image
 Input: position, size, n/m thresholds
 Output: weight
 */

double factor_DV(int position, int n, int m) {
    double weight;
    if(position<=n) {
        weight=0.0;
    }
    else {
        if(position>=m) {
            weight=1.0;
        }
        else {
            double a,b;
            a = 1.0 / (m-n);
            b = ((float)n) / (n-m);
            weight = a*position + b;
        }
    }
    return weight;
}




/******************************/
/**         access()         **/
/******************************/

/*
 Access a specific position in the array
 Input: array, dimensions, position
 Output: value at this position
 */

double access(double* array, int x, int y, int max_x) {
    return array[y*max_x+x];
}




/********************/
/**     main()     **/
/********************/

int main(int argc, char* argv[]) {
    int hor_size, sag_size, cor_size;
    int i, j;
    int h, c, s;
    int n,m;
    char *DV_ascii, *VD_ascii, *out_ascii;
    double **VD_data, **DV_data, **merged_data;
    double DV_value, VD_value, factor;
    ifstream file_in;
    ofstream file_out;
    string line;

    
    if(argc!=9) {
        cout << "*** Error. Exactly eight arguments are needed.\n";
        return 9;
    }
    
    hor_size = atoi(argv[1]);
    sag_size = atoi(argv[2]);
    cor_size = atoi(argv[3]);
    DV_ascii = argv[4];
    VD_ascii = argv[5];
    out_ascii = argv[6];
    n = atoi(argv[7]);
    m = atoi(argv[8]);
    
    printTimestamp("Ready to start.");
    
    VD_data = new double*[hor_size*cor_size];
    DV_data = new double*[hor_size*cor_size];
    merged_data = new double*[hor_size*cor_size];
    
    for(i=0;i<hor_size*cor_size;++i) {
        VD_data[i] = new double[sag_size];
        DV_data[i] = new double[sag_size];
        merged_data[i] = new double[sag_size];
    }

    
    /**************************************/
    /**     READING THE ASCII FILES      **/
    /**************************************/
    
    printTimestamp("Starting to read the input files.");

    /* VD */

    printTimestamp(VD_ascii);
    
    file_in.open(VD_ascii);
    if (not file_in.is_open()) {
        cout << "*** Error opening VD file.\n";
        return 2;
    }
    
    i=0;
    
    //    cout << "Reading VD." << endl;
    
    /* We read line by line */
    while (getline(file_in,line)) {
        if(line.length()>0) {
            istringstream sin(line);
            double val;
            int nb_values=0;
            while(sin>>val) {
                VD_data[i][nb_values]=val;
                ++nb_values;
            }
            if(nb_values!=sag_size) {
                cout << "Error? size=" << sag_size << ", but " << nb_values << " VD values read.\n";
                return -1;
            }
            ++i;
            
        }
    }
    
    file_in.close();
    
    /* DV */
    
    printTimestamp(DV_ascii);
    
    file_in.open(DV_ascii);
    if (not file_in.is_open()) {
        cout << "*** Error opening DV file.\n";
        return 2;
    }
    
    i=0;
    
    //    cout << "Reading DV." << endl;
    
    /* We read line by line */
    while (getline(file_in,line)) {
        if(line.length()>0) {
            istringstream sin(line);
            double val;
            int nb_values=0;
            while(sin>>val) {
                DV_data[i][nb_values]=val;
                ++nb_values;
            }
            if(nb_values!=sag_size) {
                cout << "Error? size=" << sag_size << ", but " << nb_values << " DV values read.\n";
                return -1;
            }
            ++i;
            
        }
    }
    
    file_in.close();
    
    
    
    /************************************/
    /**     PROCESSING THE FILES      **/
    /************************************/
    
    printTimestamp("Starting to combine the horizontal slices.");
    
    
    for(h=0; h<hor_size; ++h) {
        
        for(c=0; c<cor_size; ++c) {
            for(s=0; s<sag_size; ++s) {
                VD_value = VD_data[h*cor_size+c][s];
                DV_value = DV_data[h*cor_size+c][s];
                factor = factor_DV(h,n,m);
                merged_data[h*cor_size+c][s] = factor*DV_value + (1-factor)*VD_value;
            }
        }
        
    }
    
    
    printTimestamp("Saving the results.");

    file_out.open(out_ascii);
    if (not file_out.is_open()) {
        cout << "*** Error opening output file.\n";
        return 3;
    }
    
    for(i=0; i<hor_size*cor_size; ++i) {
        for(j=0; j<sag_size; ++j) {
            int val = (int)merged_data[i][j];
            file_out << val << " ";
        }
        file_out << endl;
    }
    file_out << endl;

    
    printTimestamp("Freeing memory.");

    
    for(i=hor_size*cor_size-1; i>=0; --i) {
        delete[] DV_data[i];
        delete[] VD_data[i];
        delete[] merged_data[i];
    }
    delete[] DV_data;
    delete[] VD_data;
    delete[] merged_data;
    
    printTimestamp("Done.");
    return 0;
}
